etcd-backup/main.go

```diff
@@ -0,0 +1,116 @@
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+	"github.com/spf13/cobra"
+	"go.etcd.io/etcd/client/pkg/v3/transport"
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"go.etcd.io/etcd/client/v3/snapshot"
+
+	"go.uber.org/zap"
+)
+
+const (
+	DefaultDialTimeout       = 15 * time.Second
+	DefaultEtcdClientTimeout = 5 * time.Minute
+)
+
+type options struct {
+	clusterID      string
+	etcdEndpoint   string
+	s3BucketName   string
+	s3BucketRegion string
+}
+
+func main() {
+	var opts options
+
+	cmd := &cobra.Command{
+		Use:          "etcd-backup",
+		SilenceUsage: true,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return run(opts)
+		},
+	}
+
+	cmd.Flags().StringVar(&opts.clusterID, "cluster-id", "", "ID of the hosted cluster, used as a key for the snapshot file in S3")
+	cmd.Flags().StringVar(&opts.etcdEndpoint, "etcd-endpoint", "", "endpoint of the etcd cluster to backup.")
+	cmd.Flags().StringVar(&opts.s3BucketName, "s3-bucket-name", "", "name of the S3 bucket to store etcd backups.")
+	cmd.Flags().StringVar(&opts.s3BucketRegion, "s3-bucket-region", "", "AWS region of the S3 bucket to store etcd backups.")
+
+	cmd.MarkFlagRequired("cluster-id")
+	cmd.MarkFlagRequired("etcd-endpoint")
+	cmd.MarkFlagRequired("s3-bucket-name")
+
+	if err := cmd.Execute(); err != nil {
+		os.Exit(1)
+	}
+}
+
+// go run etcd-backup/main.go etcd-backup --cluster-id mraee-hosted-cluster --etcd-endpoint localhost:2379 --s3-bucket-name mraee-hypershift-bucket
+func run(opts options) error {
+	tlsInfo := transport.TLSInfo{
+		CertFile:      "/etc/etcd/tls/client/etcd-client.crt",
+		KeyFile:       "/etc/etcd/tls/client/etcd-client.key",
+		TrustedCAFile: "/etc/etcd/tls/etcd-ca/ca.crt",
+	}
+
+	tlsConfig, err := tlsInfo.ClientConfig()
+	if err != nil {
+		return fmt.Errorf("error creating client TLSConfig: %w", err)
+	}
+
+	cfg := clientv3.Config{
+		Endpoints:   []string{opts.etcdEndpoint},
+		DialTimeout: DefaultDialTimeout,
+		TLS:         tlsConfig,
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), DefaultEtcdClientTimeout)
+	defer cancel()
+
+	filePath := "/tmp/snapshot.db"
+	err = snapshot.Save(ctx, zap.L(), cfg, filePath)
+	if err != nil {
+		return fmt.Errorf("failed to snapshot etcd: %w", err)
+	}
+
+	return uploadToS3(opts, filePath)
+}
+
+func uploadToS3(opts options, snapshotFilePath string) error {
+	config := aws.NewConfig()
+	// AWS_REGION must be set if s3BucketRegion is empty
+	if opts.s3BucketRegion != "" {
+		config.Region = aws.String(opts.s3BucketRegion)
+	}
+	awsSession := session.Must(session.NewSession(config))
+
+	f, err := os.Open(snapshotFilePath)
+	if err != nil {
+		return fmt.Errorf("failed to open file %q, %v", snapshotFilePath, err)
+	}
+
+	key := fmt.Sprintf("%s_%d.db", opts.clusterID, time.Now().Unix())
+	uploader := s3manager.NewUploader(awsSession, s3manager.WithUploaderRequestOptions())
+	output, err := uploader.Upload(&s3manager.UploadInput{
```

Comment:
The object should also be tagged with a minimum of `cluster_id` (internal OCM id not external OCP id) and `org_id`. @jewzaam please chime in if there is anything else we would want.

Reply:
what should the value of `org_id` be?

Reply:
I don't believe there is a spec field for it, but OCM should apply this label, `api.openshift.com/legal-entity-id` to every `HostedCluster` CR. The value of `legal-entity-id` is the organization ID in OCM.